### PR TITLE
Refactor editEntityRecord, undo, and redo to be thunks instead of generators

### PR DESCRIPTION
**Pull request by adamziel**
Builds on top of the thunks support added in #27276 and refactors just the parts of core-data required to make the `editEntityRecord`, `undo`, and `redo` work (see #28389 from @jsnajdr).

**Test plan:**

* Confirm all the tests are green.
* Go to the post editor, edit a post, add a Blog Title block and edit it as well, save everything and confirm it worked as expected.
* Try undo-ing and redo-ing adding a block, typing a few lines of text, pasting content.
* Go to the widgets editor, edit a few widgets, click Update, confirm it worked as expected.
* Try anything else edit-related that comes to your mind.

### CODE REVIEW

File: packages/core-data/src/actions.js [hunk 1]
```diff
@@ -300,44 +293,40 @@ export function* editEntityRecord( kind, name, recordId, edits, options = {} ) {
...
  */
-export function* undo() {
-	const undoEdit = yield controls.select( STORE_NAME, 'getUndoEdit' );
+export const undo = () => ( { select, dispatch } ) => {
+	const undoEdit = select.getUndoEdit();
 	if ( ! undoEdit ) {
 		return;
 	}
-	yield {
+	return dispatch( {
```
Comment (jsnajdr) [inline packages/core-data/src/actions.js]: Do we need to return the result of `dispatch` here? As the action is a plain action, the return value is going to be the action object. That doesn't seem very useful. Better to return nothing, and make the `undo` return type `void`.

File: packages/core-data/src/actions.js [hunk 2]
```diff
@@ -287,7 +280,7 @@ export function* editEntityRecord( kind, name, recordId, edits, options = {} ) {
 		}, {} ),
 		transientEdits,
 	};
-	return {
+	return await dispatch( {
```
Comment (jsnajdr) [inline packages/core-data/src/actions.js:287]: In my book, the `editEntityRecord` action is completely synchronous. It synchronously loads the entity from store, doesn't wait for resolution (quite the opposite, it throws when the entity is not already loaded), and then synchronously applies edits to the in-memory store.

**Review by jsnajdr — APPROVED** [review]
Looks good, I would just return nothing from the `undo` and `redo` actions.

Also noting that `editEntityRecord` is by nature completely sync, but that's something we'll fix later.

**Comment by fluiddot** [comment]
:wave: I noticed that after this PR got merged into `trunk` ([merge commit](https://github.com/WordPress/gutenberg/commit/8d8262c87c82acfc676bba076085b5ba28f0b0fd)), the CI check `Static Analysis (Linting, License, Type checks...)` began to fail with the following error:
```
[1] /home/runner/work/gutenberg/gutenberg/packages/core-data/src/actions.js
[1]   10:10  error  'controls' is defined but never used  no-unused-vars
```

However, I'm not sure why the CI checks passed for the last commit before merging, do you have any clue what might be causing this?

**Comment by ciampo** [comment]
This PR seems to have introduced a linting error which causes CI tasks to fail ([example](https://github.com/WordPress/gutenberg/pull/34330/checks?check_run_id=3487008405)):

```
[1] /home/runner/work/gutenberg/gutenberg/packages/core-data/src/actions.js
[1]   10:10  error  'controls' is defined but never used  no-unused-vars
```

**Comment by ciampo** [comment]
> I'm not sure why the CI checks passed for the last commit before merging

Did the `Static Analysis` task run on this PR? From a quick look, I can't see it in the list of CI checks that passed before merging

**Comment by fluiddot** [comment]
> > I'm not sure why the CI checks passed for the last commit before merging
> 
> Did the `Static Analysis` task run on this PR? From a quick look, I can't see it in the list of CI checks that passed before merging

As far as I checked, [this job run](https://github.com/WordPress/gutenberg/runs/3484619725) passed for the last commit of the branch (`3b15152`). 

Digging deeper I noticed that the job checked out a different commit 🤔 :
Screenshot 2021-09-01 at 18 53 42

**Comment by mkevins** [comment]
> > > I'm not sure why the CI checks passed for the last commit before merging
> > 
> > 
> > Did the `Static Analysis` task run on this PR? From a quick look, I can't see it in the list of CI checks that passed before merging
> 
> As far as I checked, [this job run](https://github.com/WordPress/gutenberg/runs/3484619725) passed for the last commit of the branch (`3b15152`).
> 
> Digging deeper I noticed that the job checked out a different commit  :
> Screenshot 2021-09-01 at 18 53 42

Nice digging here Carlos!

**Comment by mkevins** [comment]
Is it usually a different commit? It seems like the ref it's checking out is a merge (so maybe it's running the tests on the "would be state" of the repository if the changes from this PR are merged. I wonder if this could somehow be an issue with how git conflicts are autoresolved — perhaps subtly different outcomes when squashing or merging? I'm mentioning this since it seems this PR has two commits (https://github.com/WordPress/gutenberg/pull/34387/commits/9c0f893ae2e300d1e125a59cc504a03fe3c24525, and https://github.com/WordPress/gutenberg/pull/34387/commits/3b151527a31c15bb91bf597018d558d58756f0ea) which remove and re-add the `controls` import, respectively. When squashed, there'd be no change to that line, but I'm not sure whether an automatic conflict resolution would affect the end result after a rebase or merge :thinking:

**Comment by ciampo** [comment]
Just for completeness, a fix was merged in #34464

**Comment by adamziel** [comment]
Thank you for the prompt reaction @mkevins @fluiddot  @ciampo @ntsekouras! My best read on what happened is this:

1. All the checks passed on this PR and on https://github.com/WordPress/gutenberg/pull/34386
2. I merged https://github.com/WordPress/gutenberg/pull/34386 where some usages of `controls` were removed
3. I merged this PR, removing all the other usages of `controls`, without rebasing 

Had I rebased, the linting error would be revealed early.

**Pull request merged**